### PR TITLE
PLT-3401 Adding webserver mode to the system console.

### PR DESCRIPTION
**Pull request by crspeller**
#### Summary
- Adding webserver mode to the system console.
- Changing the default webserver mode from "regular" to "gzip"
- "regular" will now mean "gzip" so all existing servers will have gzip compression enabled upon upgrade.  

#### PMs
- Please review all strings in en.json.

### CODE REVIEW

File: webapp/components/admin_console/webserver_mode_dropdown_setting.jsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                    <td className='help-text'>
+                        <FormattedMessage
+                            id='admin.webserverModeDisabled'
+                            defaultMessage='Disabled'
+                        />
+                    </td>
+                    <td className='help-text'>
+                        <FormattedMessage
+                            id='admin.webserverModeDisabledDescription'
+                            defaultMessage='The Mattermost server will not serve static files'
```
Comment (philku79) [inline webapp/components/admin_console/webserver_mode_dropdown_setting.jsx]: Missing a "." at the end

File: webapp/i18n/en.json [hunk 2]
```diff
@@ -608,6 +608,14 @@
   "admin.user_item.switchToEmail": "Switch to Email/Password",
   "admin.user_item.sysAdmin": "System Admin",
   "admin.user_item.teamAdmin": "Team Admin",
+  "admin.webserverModeGzip": "gzip",
+  "admin.webserverModeGzipDescription": "The Mattermost server will serve static files compressed with gzip.",
+  "admin.webserverModeUncompressed": "Uncompressed",
+  "admin.webserverModeUncompressedDescription": "The Mattermost server will serve static files uncompressed.",
+  "admin.webserverModeDisabled": "Disabled",
+  "admin.webserverModeDisabledDescription": "The Mattermost server will not serve static files",
```
Comment (philku79) [inline webapp/i18n/en.json]: Missing a dot at the end like the others.

**Comment by philku79** [comment]
👍

**Comment by jwilander** [comment]
+1

File: model/config.go [hunk 3]
```diff
@@ -578,7 +578,9 @@ func (o *Config) SetDefaults() {
 
 	if o.ServiceSettings.WebserverMode == nil {
 		o.ServiceSettings.WebserverMode = new(string)
-		*o.ServiceSettings.WebserverMode = "regular"
+		*o.ServiceSettings.WebserverMode = "gzip"
+	} else if *o.ServiceSettings.WebserverMode == "regular" {
+		*o.ServiceSettings.WebserverMode = "gzip"
```
Comment (coreyhulen) [inline model/config.go]: does this always force it to gzip?  what happens if I want to disable it?

**Comment by esethna** [comment]
Hmm, test server doesn't seem to be working.

File: model/config.go (hunk 3 above)
Comment (crspeller) [inline model/config.go]: There is a new option `uncompressed` you can see that here: https://github.com/mattermost/platform/pull/3458/files#diff-469d87b06f69ef98e3dfa8a1f785570bR77

This line is so any current Mattermost server will default to using gzip.

**Comment by crspeller** [comment]
@esethna Fixed.

**Comment by esethna** [comment]
Thanks @crspeller, I took a look a the spinmint, some tweaks to consider:

1. `gzip` -> `Gzip` or `GZip` in the table and dropdown (unless lowercase is technically correct?) I see it both ways searching on google)
2. Proposed help text below the table: `Gzip compression applies to large static content files to speed up first page load. It is recommended to enable Gzip to improve performance unless your environment has specific restrictions, such as a web proxy that distributes Gzip files poorly, or if your clients are limited in processing power. This setting requires a server restart to take effect."
3. Would this setting make more sense under "File Storage" or is it more applicable in the "Configuration" section?

**Comment by crspeller** [comment]
@esethna 

1. `gzip` is the proper spelling as seen on their official site: http://www.gzip.org/ and Wikipedia page: https://en.wikipedia.org/wiki/Gzip
2. Help text updated. Edited for accuracy. Removed the line about clients with limited processing power, because I could find no evidence that this is ever the case. If a low power client really didn't want gzip compression, all is has to to is set the right header. 
3. No it would make no sense under "File Storage", "Configuration" is the most relevant section because that's where the listen address is.

**Comment by esethna** [comment]
+1, string changes look good, thanks!

**Comment by jwilander** [comment]
Just needs a rebase

**Comment by crspeller** [comment]
@jwilander Rebase complete.

**Pull request merged**